src/modules/commander/accelerometer_calibration.cpp

```diff
@@ -150,191 +154,174 @@ using namespace matrix;
 
-	unsigned errcount = 0;
+	/* use the first sensor to pace the readout, but do per-sensor counts */
+	bool done = false;
 
-	// sensor thermal corrections
-	uORB::Subscription sensor_correction_sub{ORB_ID(sensor_correction)};
-	sensor_correction_s sensor_correction{};
-	sensor_correction_sub.copy(&sensor_correction);
+	const unsigned num_accels = math::min(accel_subs.advertised_count(), MAX_ACCEL_SENS);
 
-	uORB::SubscriptionBlocking<sensor_accel_s> accel_sub[MAX_ACCEL_SENS] {
-		{ORB_ID(sensor_accel), 0, 0},
-		{ORB_ID(sensor_accel), 0, 1},
-		{ORB_ID(sensor_accel), 0, 2},
-		{ORB_ID(sensor_accel), 0, 3},
-	};
+	const hrt_abstime timestamp_start = hrt_absolute_time();
 
-	/* use the first sensor to pace the readout, but do per-sensor counts */
-	while (counts[0] < samples_num) {
-		if (accel_sub[0].updatedBlocking(100000)) {
-			for (unsigned accel_index = 0; accel_index < MAX_ACCEL_SENS; accel_index++) {
-				sensor_accel_s arp;
-
-				while (accel_sub[accel_index].update(&arp)) {
-					// fetch optional thermal offset corrections in sensor/board frame
-					Vector3f offset{0, 0, 0};
-					sensor_correction_sub.update(&sensor_correction);
-
-					if (sensor_correction.timestamp > 0 && arp.device_id != 0) {
-						for (uint8_t correction_index = 0; correction_index < MAX_ACCEL_SENS; correction_index++) {
-							if (sensor_correction.accel_device_ids[correction_index] == arp.device_id) {
-								switch (correction_index) {
-								case 0:
-									offset = Vector3f{sensor_correction.accel_offset_0};
-									break;
-								case 1:
-									offset = Vector3f{sensor_correction.accel_offset_1};
-									break;
-								case 2:
-									offset = Vector3f{sensor_correction.accel_offset_2};
-									break;
-								case 3:
-									offset = Vector3f{sensor_correction.accel_offset_3};
-									break;
-								}
-							}
-						}
-					}
+	while (!done && (hrt_elapsed_time(&timestamp_start) < 5_s)) {
+		unsigned good_count = 0;
 
-					accel_sum[accel_index] += Vector3f{arp.x, arp.y, arp.z} - offset;
+		for (unsigned accel_index = 0; accel_index < num_accels; accel_index++) {
+			sensor_accel_s arp;
 
-					counts[accel_index]++;
+			while (accel_subs[accel_index].update(&arp)) {
+				calibrations[accel_index].set_device_id(arp.device_id);
+				calibrations[accel_index].SensorCorrectionsUpdate();
 
-					if (!PX4_ISFINITE(temperature_sum[accel_index])) {
-						// set first valid value
-						temperature_sum[accel_index] = (arp.temperature * counts[accel_index]);
+				// fetch optional thermal offset corrections in sensor/board frame
+				const Vector3f accel{Vector3f{arp.x, arp.y, arp.z} - calibrations[accel_index].thermal_offset()};
 
-					} else {
-						temperature_sum[accel_index] += arp.temperature;
-					}
+				if ((accel - accel_prev[accel_index]).longerThan(0.5f)) {
+					mean[accel_index].reset();
+					accel_prev[accel_index] = accel;
+
+				} else {
+					mean[accel_index].update(accel);
```

Comment:
As data accumulation duration is short and the autopilot is supposed to be still, this may be acceptable but if the UAV turns during data collection, this mean point would not be on the sphere we want to have.
Collecting all data and letting the lm_fit do the job seems mathematically better but would use a lot more resources.

Reply:
> Collecting all data and letting the lm_fit do the job seems mathematically better but would use a lot more resources.

For reference we already run lm fit with a couple hundred points for magnetometer.

I'm in favor of keep more samples for the accel cal if we can figure out how to fit it cleanly into the user work flow. This particular logic is more about ensuring each sample is reasonably clean.

Reply:
In the actual process, I think that would not be an issue so it could be left like that.

Another solution would be to let the process as it is and use the same points that are used for the mean and just use it directly in the lm_fit.